src/main/java/uk/gov/ons/ssdc/caseprocessor/schedule/FulfilmentScheduler.java

```diff
@@ -0,0 +1,55 @@
+package uk.gov.ons.ssdc.caseprocessor.schedule;
+
+import com.godaddy.logging.Logger;
+import com.godaddy.logging.LoggerFactory;
+import java.net.InetAddress;
+import java.net.UnknownHostException;
+import java.time.OffsetDateTime;
+import java.util.Optional;
+import org.springframework.scheduling.annotation.Scheduled;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+import uk.gov.ons.ssdc.caseprocessor.model.entity.FulfilmentNextTrigger;
+import uk.gov.ons.ssdc.caseprocessor.model.repository.FulfilmentNextTriggerRepository;
+
+@Service
+public class FulfilmentScheduler {
+  private static final Logger log = LoggerFactory.getLogger(FulfilmentScheduler.class);
+  private final ClusterLeaderManager clusterLeaderManager;
+  private final FulfilmentNextTriggerRepository fulfilmentNextTriggerRepository;
+  private final FulfilmentProcessor fulfilmentProcessor;
+
+  private String hostName = InetAddress.getLocalHost().getHostName();
+
+  public FulfilmentScheduler(
+      ClusterLeaderManager clusterLeaderManager,
+      FulfilmentNextTriggerRepository fulfilmentNextTriggerRepository,
+      FulfilmentProcessor fulfilmentProcessor)
+      throws UnknownHostException {
+    this.clusterLeaderManager = clusterLeaderManager;
+    this.fulfilmentNextTriggerRepository = fulfilmentNextTriggerRepository;
+    this.fulfilmentProcessor = fulfilmentProcessor;
+  }
+
+  @Scheduled(fixedDelayString = "${scheduler.frequency}")
+  @Transactional
+  public void scheduleFulfilments() {
+    if (!clusterLeaderManager.isThisHostClusterLeader()) {
+      return; // This host (i.e. pod) is not the leader... don't do any scheduling
+    }
+
+    // Check if it's time to process the fulfilments
+    Optional<FulfilmentNextTrigger> triggerOptional =
+        fulfilmentNextTriggerRepository.findByTriggerDateTimeBefore(OffsetDateTime.now());
+
+    if (triggerOptional.isPresent()) {
+      log.with("hostName", hostName).info("Fulfilment processing triggered");
+
+      fulfilmentProcessor.addFulfilmentBatchIdAndQuantity();
+
+      FulfilmentNextTrigger trigger = triggerOptional.get();
+      trigger.setTriggerDateTime(trigger.getTriggerDateTime().plusDays(1));
```

Comment:
Possible undesirable behaviour, if you set a time some number of days in the past it would trigger back to back on repeat until it passed the current date. For instance if we scaled the service down for two days and didn't update this date before switching it back on, I think it would trigger twice immediately when scaled back up?

Reply:
yep... although presumably there'd be nothing to process the 2nd time it ran?

Also, this has desirable behaviour, if for example we wanted to have fulfilments being sent 4 times a day... it's flexible enough.

I'll have a think about whether we can put any protection in so that it errors or somesuch thing, if the date is hugely adrift... although it might cause more problems down the road

Reply:
👍 It's a nice simple and flexible design I'm happy to approve, just wanted to note some potential drawbacks that mean we'd need to be a bit careful with the dates.

Reply:
I reckon put some more thought into it in a small tech session maybe - discuss and raise another card